pkg/pillar/cmd/tpmmgr/tpmmgr.go

```diff
@@ -866,6 +873,221 @@ func getPrivateKeyFromFile(keyFile string) (*ecdsa.PrivateKey, error) {
 	return nil, err
 }
 
+func createQuoteCert() error {
+	// certificate is already created
+	if etpm.FileExists(quoteCertFile) {
+		return nil
+	}
+	// try TPM
+	if etpm.IsTpmEnabled() {
+		if err := createQuoteCertOnTpm(); err == nil {
+			return nil
+		} else {
+			// some issue with TPM. Fall back to soft cert
+			log.Errorf("createQuoteCertOnTpm failed with err (%v), trying software certificate", err)
+		}
+	}
+	// create soft certficate
+	return createQuoteCertSoft()
+}
+
+func createQuoteCertOnTpm() error {
+	//Check if we already have the certificate in /config
+	if !etpm.FileExists(quoteCertFile) {
+		//Cert is not present in /config, generate new one
+		//Store certificate in /config
+		rw, err := tpm2.OpenTPM(etpm.TpmDevicePath)
+		if err != nil {
+			return err
+		}
+		defer rw.Close()
+
+		clientCertBytes, err := ioutil.ReadFile(types.DeviceCertName)
+		if err != nil {
+			return nil
+		}
+
+		block, _ := pem.Decode(clientCertBytes)
+		if block == nil {
+			return fmt.Errorf("error in parsing clientCertBytes")
+		}
+
+		deviceCert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return err
+		}
+
+		quoteKey, _, _, err := tpm2.ReadPublic(rw, TpmQuoteKeyHdl)
+		if err != nil {
+			return err
+		}
+
+		publicKey, err := quoteKey.Key()
+		if err != nil {
+			return err
+		}
+
+		tpmPrivKey := etpm.TpmPrivateKey{}
+		tpmPrivKey.PublicKey = tpmPrivKey.Public()
+		template := createQuoteTemplate(*deviceCert)
+
+		cert, err := x509.CreateCertificate(rand.Reader,
+			&template, deviceCert, publicKey, tpmPrivKey)
+		if err != nil {
+			return err
+		}
+
+		certBlock := &pem.Block{
+			Type:  "CERTIFICATE",
+			Bytes: cert,
+		}
+
+		certBytes := pem.EncodeToMemory(certBlock)
+		if certBytes == nil {
+			return fmt.Errorf("empty bytes after encoding to PEM")
+		}
+
+		err = ioutil.WriteFile(quoteCertFile, certBytes, 0644)
+		if err != nil {
+			return err
+		}
+
+	}
+	//change state to CERTS_CREATED
+	return nil
+}
+
+// create Quote Template using the deviceCert for lifetimes
+// Use a CommonName to differentiate from the device cert itself
+func createQuoteTemplate(deviceCert x509.Certificate) x509.Certificate {
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, _ := rand.Int(rand.Reader, serialNumberLimit)
+
+	template := x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject: pkix.Name{
+			Country:      []string{"US"},
+			Province:     []string{"CA"},
+			Locality:     []string{"Santa Clara"},
+			Organization: []string{"Zededa, Inc"},
```

Comment:
No need to change this as part of this PR, but as we move the device cert generation into go code we should make the Organization etc be lf-edge and not Zededa for the generated certs in the EVE software

Reply:
Wait a second @eriknordmark @cshari-zededa -- I don't thin this is right for EVE. This probably needs to be pluggable, but the default clearly needs to belong to LF Edge organization or something similar.

Reply:
@rvs sure but we've been shipping EVE with Zededa in the Org in the certificates forever, including in the default onboarding token AFAIK.

This is just duplicating that to one more place.

Hence my request to clean this up in a separate PR.

Reply:
Well, as long as that separate PR comes quickly enough and doesn't get forgotten -- I'm OK with handling it either way @eriknordmark